<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="HybridExpress">
<title>Comparative transcriptomic analysis of hybrids and their progenitors • HybridExpress</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.8/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.8/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Comparative transcriptomic analysis of hybrids and their progenitors">
<meta property="og:description" content="HybridExpress">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">HybridExpress</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/HybridExpress.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/almeidasilvaf/HybridExpress/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Comparative transcriptomic analysis of hybrids and their progenitors</h1>
                        <h4 data-toc-skip class="author">Fabricio
Almeida-Silva</h4>
            <address class="author_afil">
      VIB-UGent Center for Plant Systems Biology, Ghent,
Belgium<br><h4 data-toc-skip class="author">Lucas
Prost-Boxoen</h4>
            <address class="author_afil">
      VIB-UGent Center for Plant Systems Biology, Ghent,
Belgium<br><h4 data-toc-skip class="author">Yves
Van de Peer</h4>
            <address class="author_afil">
      VIB-UGent Center for Plant Systems Biology, Ghent,
Belgium<br><small class="dont-index">Source: <a href="https://github.com/almeidasilvaf/HybridExpress/blob/HEAD/vignettes/HybridExpress.Rmd" class="external-link"><code>vignettes/HybridExpress.Rmd</code></a></small>
      <div class="d-none name"><code>HybridExpress.Rmd</code></div>
    </address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The formation of hybrids through the fusion of distinct genomes and
subsequent genome duplication in cases of allopolyploidy represents a
significant evolutionary event with complex effects on cellular biology,
particularly gene expression. The impact of such genome mergings and
duplications on transcription remain incompletely understood. To bridge
this gap, we introduce <strong>HybridExpress</strong>, a comprehensive
package designed to facilitate the comparative transcriptomic analysis
of hybrids and their progenitor species. <strong>HybridExpress</strong>
is tailored for RNA-Seq data derived from an ‘experimental trio’: the
hybrid organism and its two parental species. This package offers a
suite of intuitive functions enabling researchers to perform
differential expression analysis with ease, generate principal component
analysis (PCA) plots for visualizing gene expression trends, categorize
genes into 12 distinct expression pattern groups (as in <span class="citation">Rapp, Udall, and Wendel (2009)</span>), and conduct
in-depth functional analyses. Acknowledging the potential variability in
cell and transcriptome size across species and ploidy levels,
<strong>HybridExpress</strong> incorporates features for rigorous
normalization of count data. Specifically, it allows for the integration
of spike-in standards directly into the normalization process, ensuring
accurate transcriptome size adjustments when these standards are present
in the RNA-Seq count data (see full methodology in <span class="citation">Coate (2023)</span>). By offering these capabilities,
<strong>HybridExpress</strong> provides a robust toolset for unraveling
the intricate effects of genome doubling and merging on hybrid gene
expression, paving the way for novel insights into the cellular biology
of hybrid organisms.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><em><a href="https://github.com/HybridExpress" class="external-link">HybridExpress</a></em>
can be installed from Bioconductor with the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">'BiocManager'</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'BiocManager'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"HybridExpress"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load package after installation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/almeidasilvaf/HybridExpress" class="external-link">HybridExpress</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span> <span class="co"># for reproducibility</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-description">Data description<a class="anchor" aria-label="anchor" href="#data-description"></a>
</h2>
<p>For this vignette, we will use an example data set that comprises
(unpublished) gene expression data (in counts) for <em>Chlamydomonas
reinhardtii</em>. In our lab, we crossed a diploid line of <em>C.
reinhardtii</em> (hereafter “P1”) with a haploid line (hereafter “P2”),
thus generating a triploid line through the merging of the two parental
genomes. The count matrix and sample metadata are stored in
<code>SummarizedExperiment</code> objects, which is the standard
Bioconductor data structure required by <em><a href="https://github.com/HybridExpress" class="external-link">HybridExpress</a></em>. For
instructions on how to create a <code>SummarizedExperiment</code>
object, check the FAQ section of this vignette.</p>
<p>Let’s load the example data and take a quick look at it:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">se_chlamy</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect the `SummarizedExperiment` object</span></span>
<span><span class="va">se_chlamy</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 13058 18 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(13058): Cre01.g000050 Cre01.g000150 ... ERCC-00170 ERCC-00171</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(18): S1 S2 ... S17 S18</span></span>
<span><span class="co">#&gt; colData names(2): Ploidy Generation</span></span>
<span></span>
<span><span class="co">## Take a look at the colData and count matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se_chlamy</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 18 rows and 2 columns</span></span>
<span><span class="co">#&gt;          Ploidy Generation</span></span>
<span><span class="co">#&gt;     &lt;character&gt;   &lt;factor&gt;</span></span>
<span><span class="co">#&gt; S1      diploid         P1</span></span>
<span><span class="co">#&gt; S2      diploid         P1</span></span>
<span><span class="co">#&gt; S3      diploid         P1</span></span>
<span><span class="co">#&gt; S4      diploid         P1</span></span>
<span><span class="co">#&gt; S5      diploid         P1</span></span>
<span><span class="co">#&gt; ...         ...        ...</span></span>
<span><span class="co">#&gt; S14    triploid         F1</span></span>
<span><span class="co">#&gt; S15    triploid         F1</span></span>
<span><span class="co">#&gt; S16    triploid         F1</span></span>
<span><span class="co">#&gt; S17    triploid         F1</span></span>
<span><span class="co">#&gt; S18    triploid         F1</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_chlamy</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 S1  S2   S3   S4  S5   S6   S7   S8   S9 S10 S11  S12  S13  S14</span></span>
<span><span class="co">#&gt; Cre01.g000050   31  21   26   33  19   48   17    6   13   6  10    7   31   22</span></span>
<span><span class="co">#&gt; Cre01.g000150   50  29   35   99  11   58   51   46   41  33  28   14   53   25</span></span>
<span><span class="co">#&gt; Cre01.g000200   36  26   24   29  17   36   14   15   12   8  14    7   25   24</span></span>
<span><span class="co">#&gt; Cre01.g000250  440 272  394  332 283  585  272  274  255 160 225  235  405  391</span></span>
<span><span class="co">#&gt; Cre01.g000300 1242 839 1216 1251 811 1785 1341 1306 1122 877 844 1082 1704 1739</span></span>
<span><span class="co">#&gt; Cre01.g000350  412 264  294  336 252  478  233  221  195 155 190  201  299  272</span></span>
<span><span class="co">#&gt;                S15  S16  S17  S18</span></span>
<span><span class="co">#&gt; Cre01.g000050   29   22   21   16</span></span>
<span><span class="co">#&gt; Cre01.g000150   37   17   24   21</span></span>
<span><span class="co">#&gt; Cre01.g000200   24   26   18   21</span></span>
<span><span class="co">#&gt; Cre01.g000250  358  339  340  332</span></span>
<span><span class="co">#&gt; Cre01.g000300 1524 1720 1517 1243</span></span>
<span><span class="co">#&gt; Cre01.g000350  276  324  246  275</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">se_chlamy</span><span class="op">$</span><span class="va">Ploidy</span>, <span class="va">se_chlamy</span><span class="op">$</span><span class="va">Generation</span><span class="op">)</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt;            F1 P1 P2</span></span>
<span><span class="co">#&gt;   diploid   0  6  0</span></span>
<span><span class="co">#&gt;   haploid   0  0  6</span></span>
<span><span class="co">#&gt;   triploid  6  0  0</span></span></code></pre></div>
<p>As you can see, the count matrix contains 13058 genes and 18 samples,
with 6 replicates for parent 1 (P1, diploid), 6 replicates for parent 2
(P2, haploid), and 6 replicates for the progeny (F1, triploid).</p>
</div>
<div class="section level2">
<h2 id="adding-midparent-expression-values">Adding midparent expression values<a class="anchor" aria-label="anchor" href="#adding-midparent-expression-values"></a>
</h2>
<p>First of all, you’d want to add in your count matrix <em>in
silico</em> samples that contain the expression values of the midparent.
This can be done with the function
<code><a href="../reference/add_midparent_expression.html">add_midparent_expression()</a></code>, which takes a random sample
pair (one sample from each parent, sampling without replacement) and
calculates the midparent expression value in one of three ways:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Mean (default):</strong> get the mean expression of the two
samples.</li>
<li>
<strong>Sum:</strong> get the sum of the two samples.</li>
<li>
<strong>Weighted mean:</strong> get the weighted mean of the two
samples by multiplying the expression value of each parent by a weight.
Typically, this can be used if the two parents have different ploidy
levels, and the weights would correspond to the ploidy level of each
parent.</li>
</ol>
<p>For this function, besides specifying the method to obtain the
midparent expression values (i.e., “mean”, “sum”, or “weightedmean”),
users must also specify the name of the column in <code>colData</code>
that contains information about the generations (default:
<em>“Generation”</em>), as well as the levels corresponding to each
parent (default: <em>“P1”</em> and <em>“P2”</em> for parents 1 and 2,
respectively).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add midparent expression using the mean of the counts</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_midparent_expression.html">add_midparent_expression</a></span><span class="op">(</span><span class="va">se_chlamy</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 S1  S2   S3   S4  S5   S6   S7   S8   S9 S10 S11  S12  S13  S14</span></span>
<span><span class="co">#&gt; Cre01.g000050   31  21   26   33  19   48   17    6   13   6  10    7   31   22</span></span>
<span><span class="co">#&gt; Cre01.g000150   50  29   35   99  11   58   51   46   41  33  28   14   53   25</span></span>
<span><span class="co">#&gt; Cre01.g000200   36  26   24   29  17   36   14   15   12   8  14    7   25   24</span></span>
<span><span class="co">#&gt; Cre01.g000250  440 272  394  332 283  585  272  274  255 160 225  235  405  391</span></span>
<span><span class="co">#&gt; Cre01.g000300 1242 839 1216 1251 811 1785 1341 1306 1122 877 844 1082 1704 1739</span></span>
<span><span class="co">#&gt; Cre01.g000350  412 264  294  336 252  478  233  221  195 155 190  201  299  272</span></span>
<span><span class="co">#&gt;                S15  S16  S17  S18 midparent1 midparent2 midparent3 midparent4</span></span>
<span><span class="co">#&gt; Cre01.g000050   29   22   21   16         18         27         14         20</span></span>
<span><span class="co">#&gt; Cre01.g000150   37   17   24   21         32         46         38         56</span></span>
<span><span class="co">#&gt; Cre01.g000200   24   26   18   21         19         22         20         18</span></span>
<span><span class="co">#&gt; Cre01.g000250  358  339  340  332        310        372        273        284</span></span>
<span><span class="co">#&gt; Cre01.g000300 1524 1720 1517 1243       1030       1331       1072       1166</span></span>
<span><span class="co">#&gt; Cre01.g000350  276  324  246  275        242        316        242        268</span></span>
<span><span class="co">#&gt;               midparent5 midparent6</span></span>
<span><span class="co">#&gt; Cre01.g000050         18         22</span></span>
<span><span class="co">#&gt; Cre01.g000150         31         46</span></span>
<span><span class="co">#&gt; Cre01.g000200         16         24</span></span>
<span><span class="co">#&gt; Cre01.g000250        278        348</span></span>
<span><span class="co">#&gt; Cre01.g000300       1076       1182</span></span>
<span><span class="co">#&gt; Cre01.g000350        242        304</span></span>
<span></span>
<span><span class="co"># Alternative 1: using the sum of the counts</span></span>
<span><span class="fu"><a href="../reference/add_midparent_expression.html">add_midparent_expression</a></span><span class="op">(</span><span class="va">se_chlamy</span>, method <span class="op">=</span> <span class="st">"sum"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 S1  S2   S3   S4  S5   S6   S7   S8   S9 S10 S11  S12  S13  S14</span></span>
<span><span class="co">#&gt; Cre01.g000050   31  21   26   33  19   48   17    6   13   6  10    7   31   22</span></span>
<span><span class="co">#&gt; Cre01.g000150   50  29   35   99  11   58   51   46   41  33  28   14   53   25</span></span>
<span><span class="co">#&gt; Cre01.g000200   36  26   24   29  17   36   14   15   12   8  14    7   25   24</span></span>
<span><span class="co">#&gt; Cre01.g000250  440 272  394  332 283  585  272  274  255 160 225  235  405  391</span></span>
<span><span class="co">#&gt; Cre01.g000300 1242 839 1216 1251 811 1785 1341 1306 1122 877 844 1082 1704 1739</span></span>
<span><span class="co">#&gt; Cre01.g000350  412 264  294  336 252  478  233  221  195 155 190  201  299  272</span></span>
<span><span class="co">#&gt;                S15  S16  S17  S18 midparent1 midparent2 midparent3 midparent4</span></span>
<span><span class="co">#&gt; Cre01.g000050   29   22   21   16         43         26         58         46</span></span>
<span><span class="co">#&gt; Cre01.g000150   37   17   24   21         86         25         86        140</span></span>
<span><span class="co">#&gt; Cre01.g000200   24   26   18   21         38         24         50         41</span></span>
<span><span class="co">#&gt; Cre01.g000250  358  339  340  332        666        518        810        587</span></span>
<span><span class="co">#&gt; Cre01.g000300 1524 1720 1517 1243       2557       1893       2629       2373</span></span>
<span><span class="co">#&gt; Cre01.g000350  276  324  246  275        527        453        668        531</span></span>
<span><span class="co">#&gt;               midparent5 midparent6</span></span>
<span><span class="co">#&gt; Cre01.g000050         37         27</span></span>
<span><span class="co">#&gt; Cre01.g000150         96         62</span></span>
<span><span class="co">#&gt; Cre01.g000200         51         34</span></span>
<span><span class="co">#&gt; Cre01.g000250        714        432</span></span>
<span><span class="co">#&gt; Cre01.g000300       2548       1716</span></span>
<span><span class="co">#&gt; Cre01.g000350        633        419</span></span>
<span></span>
<span><span class="co"># Alternative 2: using the weighted mean of the counts (weights = ploidy)</span></span>
<span><span class="va">w</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span> <span class="co"># P1 = diploid; P2 = haploid</span></span>
<span><span class="fu"><a href="../reference/add_midparent_expression.html">add_midparent_expression</a></span><span class="op">(</span><span class="va">se_chlamy</span>, method <span class="op">=</span> <span class="st">"weightedmean"</span>, weights <span class="op">=</span> <span class="va">w</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 S1  S2   S3   S4  S5   S6   S7   S8   S9 S10 S11  S12  S13  S14</span></span>
<span><span class="co">#&gt; Cre01.g000050   31  21   26   33  19   48   17    6   13   6  10    7   31   22</span></span>
<span><span class="co">#&gt; Cre01.g000150   50  29   35   99  11   58   51   46   41  33  28   14   53   25</span></span>
<span><span class="co">#&gt; Cre01.g000200   36  26   24   29  17   36   14   15   12   8  14    7   25   24</span></span>
<span><span class="co">#&gt; Cre01.g000250  440 272  394  332 283  585  272  274  255 160 225  235  405  391</span></span>
<span><span class="co">#&gt; Cre01.g000300 1242 839 1216 1251 811 1785 1341 1306 1122 877 844 1082 1704 1739</span></span>
<span><span class="co">#&gt; Cre01.g000350  412 264  294  336 252  478  233  221  195 155 190  201  299  272</span></span>
<span><span class="co">#&gt;                S15  S16  S17  S18 midparent1 midparent2 midparent3 midparent4</span></span>
<span><span class="co">#&gt; Cre01.g000050   29   22   21   16         25         29         39         22</span></span>
<span><span class="co">#&gt; Cre01.g000150   37   17   24   21         27         30         29         16</span></span>
<span><span class="co">#&gt; Cre01.g000200   24   26   18   21         27         32         33         21</span></span>
<span><span class="co">#&gt; Cre01.g000250  358  339  340  332        181        232        270        210</span></span>
<span><span class="co">#&gt; Cre01.g000300 1524 1720 1517 1243       1453       1576       1753       1532</span></span>
<span><span class="co">#&gt; Cre01.g000350  276  324  246  275        166        202        223        165</span></span>
<span><span class="co">#&gt;               midparent5 midparent6</span></span>
<span><span class="co">#&gt; Cre01.g000050         26         17</span></span>
<span><span class="co">#&gt; Cre01.g000150         48         15</span></span>
<span><span class="co">#&gt; Cre01.g000200         29         17</span></span>
<span><span class="co">#&gt; Cre01.g000250        202        148</span></span>
<span><span class="co">#&gt; Cre01.g000300       1705       1125</span></span>
<span><span class="co">#&gt; Cre01.g000350        186        136</span></span></code></pre></div>
<p>We will proceed our analyses with the midparent expression values
obtained from the mean of the counts, stored in the <code>se</code>
object.</p>
</div>
<div class="section level2">
<h2 id="normalizing-count-data">Normalizing count data<a class="anchor" aria-label="anchor" href="#normalizing-count-data"></a>
</h2>
<p>To normalize count data, the function <code><a href="../reference/add_size_factors.html">add_size_factors()</a></code>
calculates size factors (used by <em><a href="https://bioconductor.org/packages/3.19/DESeq2" class="external-link">DESeq2</a></em> for
normalization) and adds them as an extra column in the colData slot of
your <code>SummarizedExperiment</code> object. Such size factors are
calculated using one of two methods:</p>
<ol style="list-style-type: decimal">
<li><p>By library size (default) using the ‘median of ratios’ method
implemented in <em><a href="https://bioconductor.org/packages/3.19/DESeq2" class="external-link">DESeq2</a></em>.</p></li>
<li><p>By cell size/biomass using spike-in controls (if available). If
spike-in controls are present in the count matrix, you can use them for
normalization by setting <code>spikein = TRUE</code> and specifying the
pattern used to indicate rows that contain spike-ins (usually they start
with <em>ERCC</em>)<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;strong&gt;Note:&lt;/strong&gt; if you use spike-in
normalization, the function &lt;code&gt;&lt;a href="../reference/add_size_factors.html"&gt;add_size_factors()&lt;/a&gt;&lt;/code&gt;
automatically removes rows containing spike-in counts from the
&lt;code&gt;SummarizedExperiment&lt;/code&gt; object after normalization. However,
if you have counts for spike-in controls in your count matrix, but do
not want to use spike-in normalization, you should remove them before
creating the &lt;code&gt;SummarizedExperiment&lt;/code&gt; object.&lt;/p&gt;'><sup>1</sup></a>. Normalization with spike-in controls is
particularly useful if the amount of mRNA per cell is not equal between
generations (due to, for instance, different ploidy levels, which in
turn can lead to different cell sizes and/or biomass).</p></li>
</ol>
<p>In our example data set, spike-in controls are available in the last
rows of the count matrix. Let’s take a look at them.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Show last rows of the count matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;              S1   S2   S3   S4   S5   S6   S7   S8   S9  S10  S11  S12  S13</span></span>
<span><span class="co">#&gt; ERCC-00163   74   75   55   77   51   84  132  127   93  108   79  102   66</span></span>
<span><span class="co">#&gt; ERCC-00164    0    4    0    1    4    1    6    4    1    4    3    2    5</span></span>
<span><span class="co">#&gt; ERCC-00165  147  139   87  165  118  179  236  246  139  218  176  145  118</span></span>
<span><span class="co">#&gt; ERCC-00168    2    5    2    2    2    6    5    1    1    5    0    3    4</span></span>
<span><span class="co">#&gt; ERCC-00170   97   95   73  101   70  118  186  148  110  167  110  103   72</span></span>
<span><span class="co">#&gt; ERCC-00171 4644 4959 3554 5357 4170 5946 8207 7915 4992 7843 6455 5884 4537</span></span>
<span><span class="co">#&gt;             S14  S15  S16  S17  S18 midparent1 midparent2 midparent3 midparent4</span></span>
<span><span class="co">#&gt; ERCC-00163   67   62   80   88   47         67         96        101         90</span></span>
<span><span class="co">#&gt; ERCC-00164    4    3    5    4    2          2          2          4          2</span></span>
<span><span class="co">#&gt; ERCC-00165  143  115  185  192  136        132        198        192        155</span></span>
<span><span class="co">#&gt; ERCC-00168    6    2   11    5    2          1          6          3          2</span></span>
<span><span class="co">#&gt; ERCC-00170   99   64  121  103   84         92        142        122        102</span></span>
<span><span class="co">#&gt; ERCC-00171 5665 4598 6168 5771 4747       5004       6894       6437       5620</span></span>
<span><span class="co">#&gt;            midparent5 midparent6</span></span>
<span><span class="co">#&gt; ERCC-00163         92         84</span></span>
<span><span class="co">#&gt; ERCC-00164          5          0</span></span>
<span><span class="co">#&gt; ERCC-00165        177        143</span></span>
<span><span class="co">#&gt; ERCC-00168          4          2</span></span>
<span><span class="co">#&gt; ERCC-00170        128        104</span></span>
<span><span class="co">#&gt; ERCC-00171       6188       4818</span></span></code></pre></div>
<p>As you can see, rows with spike-in counts start with “ERCC”. Let’s
add size factors to our <code>SummarizedExperiment</code> object using
spike-in controls.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Take a look at the original colData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 24 rows and 2 columns</span></span>
<span><span class="co">#&gt;                 Ploidy Generation</span></span>
<span><span class="co">#&gt;            &lt;character&gt;   &lt;factor&gt;</span></span>
<span><span class="co">#&gt; S1             diploid         P1</span></span>
<span><span class="co">#&gt; S2             diploid         P1</span></span>
<span><span class="co">#&gt; S3             diploid         P1</span></span>
<span><span class="co">#&gt; S4             diploid         P1</span></span>
<span><span class="co">#&gt; S5             diploid         P1</span></span>
<span><span class="co">#&gt; ...                ...        ...</span></span>
<span><span class="co">#&gt; midparent2          NA  midparent</span></span>
<span><span class="co">#&gt; midparent3          NA  midparent</span></span>
<span><span class="co">#&gt; midparent4          NA  midparent</span></span>
<span><span class="co">#&gt; midparent5          NA  midparent</span></span>
<span><span class="co">#&gt; midparent6          NA  midparent</span></span>
<span></span>
<span><span class="co"># Add size factors estimated from spike-in controls</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_size_factors.html">add_size_factors</a></span><span class="op">(</span><span class="va">se</span>, spikein <span class="op">=</span> <span class="cn">TRUE</span>, spikein_pattern <span class="op">=</span> <span class="st">"ERCC"</span><span class="op">)</span></span>
<span><span class="co">#&gt; converting counts to integer mode</span></span>
<span></span>
<span><span class="co"># Take a look at the new colData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 24 rows and 3 columns</span></span>
<span><span class="co">#&gt;                 Ploidy Generation sizeFactor</span></span>
<span><span class="co">#&gt;            &lt;character&gt;   &lt;factor&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; S1             diploid         P1   0.856262</span></span>
<span><span class="co">#&gt; S2             diploid         P1   0.938417</span></span>
<span><span class="co">#&gt; S3             diploid         P1   0.648653</span></span>
<span><span class="co">#&gt; S4             diploid         P1   0.969369</span></span>
<span><span class="co">#&gt; S5             diploid         P1   0.718761</span></span>
<span><span class="co">#&gt; ...                ...        ...        ...</span></span>
<span><span class="co">#&gt; midparent2          NA  midparent   1.278078</span></span>
<span><span class="co">#&gt; midparent3          NA  midparent   1.201072</span></span>
<span><span class="co">#&gt; midparent4          NA  midparent   1.000892</span></span>
<span><span class="co">#&gt; midparent5          NA  midparent   1.186773</span></span>
<span><span class="co">#&gt; midparent6          NA  midparent   0.871889</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="exploratory-data-analyses">Exploratory data analyses<a class="anchor" aria-label="anchor" href="#exploratory-data-analyses"></a>
</h2>
<p>Next, you can perform exploratory analyses of sample clustering to
verify if samples group as expected. With <em><a href="https://github.com/HybridExpress" class="external-link">HybridExpress</a></em>, this can
be performed using two functions:</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="../reference/pca_plot.html">pca_plot()</a></code>: creates principal component analysis (PCA)
plots, with colors and shapes (optional) mapped to levels of
<code>colData</code> variables;</li>
<li>
<code><a href="../reference/plot_samplecor.html">plot_samplecor()</a></code>: plots a heatmap of hierarchically
clustered pairwise sample correlations.</li>
</ol>
<p>Let’s start with the PCA plot:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For colData rows with missing values (midparent samples), add "midparent"</span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">Ploidy</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">Ploidy</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"midparent"</span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">Generation</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">Generation</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"midparent"</span></span>
<span></span>
<span><span class="co"># Create PCA plot</span></span>
<span><span class="fu"><a href="../reference/pca_plot.html">pca_plot</a></span><span class="op">(</span><span class="va">se</span>, color_by <span class="op">=</span> <span class="st">"Generation"</span>, shape_by <span class="op">=</span> <span class="st">"Ploidy"</span>, add_mean <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="HybridExpress_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>In the plot above, each data point corresponds to a sample, and
colors and shapes are mapped to levels of the variables specified in the
arguments <code>color_by</code> and <code>shape_by</code>, respectively.
Besides, by specifying <code>add_mean = TRUE</code>, we added a diamond
shape indicating the mean PC coordinates based on the variable in
<code>color_by</code> (here, <em>“Generation”</em>).</p>
<p>Now, let’s plot the heatmap of sample correlations:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot a heatmap of sample correlations</span></span>
<span><span class="fu"><a href="../reference/plot_samplecor.html">plot_samplecor</a></span><span class="op">(</span><span class="va">se</span>, coldata_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Generation"</span>, <span class="st">"Ploidy"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="HybridExpress_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>We can see that samples group well together both in the PCA plot and
in the correlation heatmap.</p>
<p>Of note, both <code><a href="../reference/pca_plot.html">pca_plot()</a></code> and
<code><a href="../reference/plot_samplecor.html">plot_samplecor()</a></code> use only the top 500 genes with the
highest variances to create the plot. This is because genes with low
variances (i.e., genes that do not vary much across samples) are
uninformative and typically only add noise. You can change this number
(to use more or less genes) in the <code>ntop</code> argument of both
functions.</p>
</div>
<div class="section level2">
<h2 id="identifying-differentially-expressed-genes-between-hybrids-and-their-parents">Identifying differentially expressed genes between hybrids and their
parents<a class="anchor" aria-label="anchor" href="#identifying-differentially-expressed-genes-between-hybrids-and-their-parents"></a>
</h2>
<p>To compare gene expression levels of hybrids to their progenitor
species, you can use the function <code><a href="../reference/get_deg_list.html">get_deg_list()</a></code>. This
function performs differential expression analyses using <em><a href="https://bioconductor.org/packages/3.19/DESeq2" class="external-link">DESeq2</a></em> and
returns a list of data frames with gene-wise test statistics for the
following contrasts:</p>
<ol style="list-style-type: decimal">
<li>
<code>P2_vs_P1</code>: parent 2 (numerator) versus parent 1
(denominator).</li>
<li>
<code>F1_vs_P1</code>: hybrid (numerator) versus parent 1
(denominator).</li>
<li>
<code>F1_vs_P2</code>: hybrid (numerator) versus parent 2
(denominator).</li>
<li>
<code>F1_vs_midparent</code>: hybrid (numerator) vs midparent
(denominator).</li>
</ol>
<p>The size factors estimated with <code><a href="../reference/add_size_factors.html">add_size_factors()</a></code> are
used for normalization before differential expression testing. Let’s use
<code><a href="../reference/get_deg_list.html">get_deg_list()</a></code> to get differentially expressed genes (DEGs)
for each contrast.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a list of differentially expressed genes for each contrast</span></span>
<span><span class="va">deg_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_deg_list.html">get_deg_list</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="co">#&gt; using pre-existing size factors</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span>
<span></span>
<span><span class="co"># Inspecting the output</span></span>
<span><span class="co">## Getting contrast names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">deg_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "P2_vs_P1"        "F1_vs_P1"        "F1_vs_P2"        "F1_vs_midparent"</span></span>
<span></span>
<span><span class="co">## Accessing gene-wise test statistics for contrast `F1_vs_P1`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">deg_list</span><span class="op">$</span><span class="va">F1_vs_P1</span><span class="op">)</span></span>
<span><span class="co">#&gt;                baseMean log2FoldChange     lfcSE      stat       pvalue</span></span>
<span><span class="co">#&gt; Cre01.g000450  39.59273     -1.1632346 0.2165407 -5.371898 7.791211e-08</span></span>
<span><span class="co">#&gt; Cre01.g000750 103.10296      0.8807307 0.1842856  4.779162 1.760270e-06</span></span>
<span><span class="co">#&gt; Cre01.g000900 471.44084      0.6807925 0.1980730  3.437079 5.880239e-04</span></span>
<span><span class="co">#&gt; Cre01.g001000  10.34574     -6.6304915 0.9631384 -6.884256 5.809030e-12</span></span>
<span><span class="co">#&gt; Cre01.g001100 453.87151     -0.6457115 0.1506903 -4.285025 1.827187e-05</span></span>
<span><span class="co">#&gt; Cre01.g001200 159.03643      0.7231060 0.1946090  3.715685 2.026536e-04</span></span>
<span><span class="co">#&gt;                       padj</span></span>
<span><span class="co">#&gt; Cre01.g000450 4.653531e-07</span></span>
<span><span class="co">#&gt; Cre01.g000750 8.329159e-06</span></span>
<span><span class="co">#&gt; Cre01.g000900 1.707442e-03</span></span>
<span><span class="co">#&gt; Cre01.g001000 6.659069e-11</span></span>
<span><span class="co">#&gt; Cre01.g001100 7.200448e-05</span></span>
<span><span class="co">#&gt; Cre01.g001200 6.453951e-04</span></span>
<span></span>
<span><span class="co">## Counting the number of DEGs per contrast</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">deg_list</span>, <span class="va">nrow</span><span class="op">)</span></span>
<span><span class="co">#&gt;        P2_vs_P1        F1_vs_P1        F1_vs_P2 F1_vs_midparent </span></span>
<span><span class="co">#&gt;            8698            5476            7350            4348</span></span></code></pre></div>
<p>To summarize the frequencies of up- and down-regulated genes per
contrast in a single data frame, use the function
<code><a href="../reference/get_deg_counts.html">get_deg_counts()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a data frame with DEG frequencies for each contrast</span></span>
<span><span class="va">deg_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_deg_counts.html">get_deg_counts</a></span><span class="op">(</span><span class="va">deg_list</span><span class="op">)</span></span>
<span></span>
<span><span class="va">deg_counts</span></span>
<span><span class="co">#&gt;          contrast   up down total perc_up perc_down perc_total</span></span>
<span><span class="co">#&gt; 1        P2_vs_P1  828 7870  8698     6.4      60.7       67.1</span></span>
<span><span class="co">#&gt; 2        F1_vs_P1 1471 4005  5476    11.3      30.9       42.2</span></span>
<span><span class="co">#&gt; 3        F1_vs_P2 6487  863  7350    50.0       6.7       56.7</span></span>
<span><span class="co">#&gt; 4 F1_vs_midparent 2620 1728  4348    20.2      13.3       33.5</span></span></code></pre></div>
<p>It is important to note that the columns <code>perc_up</code>,
<code>perc_down</code>, and <code>perc_total</code> show the percentages
of up-regulated, down-regulated, and all differentially expressed genes
relative to the total number of genes in the count matrix. The total
number of genes in the count matrix is stored in the <code>ngenes</code>
attribute of the list returned by <code><a href="../reference/get_deg_list.html">get_deg_list()</a></code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">deg_list</span><span class="op">)</span><span class="op">$</span><span class="va">ngenes</span></span>
<span><span class="co">#&gt; [1] 12966</span></span></code></pre></div>
<p>However, since the count matrix usually does not include all genes in
the genome (e.g., lowly expressed genes and genes with low variance are
usually filtered out), the percentages in <code>perc_up</code>,
<code>perc_down</code>, and <code>perc_total</code> are not relative to
the total number of genes in the genome. To use the total number of
genes in the genome as the reference, we need to update the
<code>ngenes</code> attribute of the DEG list with the appropriate
number as follows:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Total number of genes in the C. reinhardtii genome (v6.1): 16883</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">deg_list</span><span class="op">)</span><span class="op">$</span><span class="va">ngenes</span> <span class="op">&lt;-</span> <span class="fl">16883</span></span></code></pre></div>
<p>Then, we can run <code><a href="../reference/get_deg_counts.html">get_deg_counts()</a></code> again to get the
percentages relative to the total number of genes in the genome.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">deg_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_deg_counts.html">get_deg_counts</a></span><span class="op">(</span><span class="va">deg_list</span><span class="op">)</span></span>
<span><span class="va">deg_counts</span></span>
<span><span class="co">#&gt;          contrast   up down total perc_up perc_down perc_total</span></span>
<span><span class="co">#&gt; 1        P2_vs_P1  828 7870  8698     4.9      46.6       51.5</span></span>
<span><span class="co">#&gt; 2        F1_vs_P1 1471 4005  5476     8.7      23.7       32.4</span></span>
<span><span class="co">#&gt; 3        F1_vs_P2 6487  863  7350    38.4       5.1       43.5</span></span>
<span><span class="co">#&gt; 4 F1_vs_midparent 2620 1728  4348    15.5      10.2       25.8</span></span></code></pre></div>
<p>Finally, we can summarize everything in a single publication-ready
figure using the plot <code><a href="../reference/plot_expression_triangle.html">plot_expression_triangle()</a></code>, which
shows the ‘experimental trio’ (i.e., hybrid and its progenitors) as a
triangle, with the frequencies of DEGs indicated.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot expression triangle</span></span>
<span><span class="fu"><a href="../reference/plot_expression_triangle.html">plot_expression_triangle</a></span><span class="op">(</span><span class="va">deg_counts</span><span class="op">)</span></span></code></pre></div>
<p><img src="HybridExpress_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>This figure is commonly used in publications, and it was inspired by
<span class="citation">Rapp, Udall, and Wendel (2009)</span>. For each
edge (line), numbers in the middle (in bold) indicate the frequency of
all DEGs, and numbers at the ends (close to boxes) indicate the
frequency of up-regulated genes in each generation. For instance, the
figure above shows that, for the contrast between F1 and P1, there are
5476 DEGs (32.4% of the genome), of which 1471 are up-regulated in F1,
and 4005 are up-regulated in P1.</p>
<p>For a custom figure, you can also specify your own color palette and
labels for the boxes. For example:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create vectors (length 4) of colors and box labels</span></span>
<span><span class="va">pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"springgreen4"</span>, <span class="st">"darkorange3"</span>, <span class="st">"mediumpurple4"</span>, <span class="st">"mediumpurple3"</span><span class="op">)</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Parent 1\n(2n)"</span>, <span class="st">"Parent 2\n(n)"</span>, <span class="st">"Progeny\n(3n)"</span>, <span class="st">"Midparent"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_expression_triangle.html">plot_expression_triangle</a></span><span class="op">(</span><span class="va">deg_counts</span>, palette <span class="op">=</span> <span class="va">pal</span>, box_labels <span class="op">=</span> <span class="va">labels</span><span class="op">)</span></span></code></pre></div>
<p><img src="HybridExpress_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="expression-based-gene-classification">Expression-based gene classification<a class="anchor" aria-label="anchor" href="#expression-based-gene-classification"></a>
</h2>
<p>After identifying DEGs for different contrasts, you’d typically want
to classify your genes into expression partitions based on their
expression patterns. This can be performed with the function
<code><a href="../reference/expression_partitioning.html">expression_partitioning()</a></code>, which classifies genes into one
of the 12 <strong>categories</strong> as in <span class="citation">Rapp,
Udall, and Wendel (2009)</span>, and into 5 major
<strong>classes</strong> that summarize the 12 categories. The five
classes are:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Transgressive up-regulation (UP):</strong> gene is
up-regulated in the hybrid as compared to both parents.</li>
<li>
<strong>Transgressive down-regulation (DOWN):</strong> gene is
down-regulated in the hybrid as compared to both parents.</li>
<li>
<strong>Additivity (ADD):</strong> gene expression in the hybrid is
the mean of both parents (additive effect).</li>
<li>
<strong>Expression-level dominance toward parent 1
(ELD_P1):</strong> gene expression in the hybrid is the same as in
parent 1, but different from parent 2.</li>
<li>
<strong>Expression-level dominance toward parent 2
(ELD_P2):</strong> gene expression in the hybrid is the same as in
parent 2, but different from parent 1.</li>
</ol>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Classify genes in expression partitions</span></span>
<span><span class="va">exp_partitions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/expression_partitioning.html">expression_partitioning</a></span><span class="op">(</span><span class="va">deg_list</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect the output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">exp_partitions</span><span class="op">)</span></span>
<span><span class="co">#&gt;            Gene Category Class lFC_F1_vs_P1 lFC_F1_vs_P2</span></span>
<span><span class="co">#&gt; 1 Cre01.g003650        1   ADD    0.7838125   -1.0484464</span></span>
<span><span class="co">#&gt; 2 Cre01.g005150        1   ADD    1.0473362   -0.5041601</span></span>
<span><span class="co">#&gt; 3 Cre01.g008600        1   ADD    5.0518384   -1.4840829</span></span>
<span><span class="co">#&gt; 4 Cre01.g013500        1   ADD    2.1099265   -1.5329846</span></span>
<span><span class="co">#&gt; 5 Cre01.g034850        1   ADD    1.5838851   -0.7611868</span></span>
<span><span class="co">#&gt; 6 Cre01.g800005        1   ADD    1.4928449   -0.9315119</span></span>
<span></span>
<span><span class="co"># Count number of genes per category</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">exp_partitions</span><span class="op">$</span><span class="va">Category</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    1    2    3    4    5    6    7    8    9   10   11   12 </span></span>
<span><span class="co">#&gt;   70  262   66 3283   77  287  214  651  280  147 1760 1666</span></span>
<span></span>
<span><span class="co"># Count number of genes per class</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">exp_partitions</span><span class="op">$</span><span class="va">Class</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     UP   DOWN    ADD ELD_P1 ELD_P2 </span></span>
<span><span class="co">#&gt;   1015    427   1736   3563   2022</span></span></code></pre></div>
<p>To visualize the expression partitions as a scatter plot of
expression divergences, you can use the function
<code><a href="../reference/plot_expression_partitions.html">plot_expression_partitions()</a></code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot partitions as a scatter plot of divergences</span></span>
<span><span class="fu"><a href="../reference/plot_expression_partitions.html">plot_expression_partitions</a></span><span class="op">(</span><span class="va">exp_partitions</span>, group_by <span class="op">=</span> <span class="st">"Category"</span><span class="op">)</span></span></code></pre></div>
<p><img src="HybridExpress_files/figure-html/unnamed-chunk-15-1.png" width="768"></p>
<p>By default, genes are grouped by <code>Category</code>. However, you
can also group genes by <code>Class</code> as follows:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Group by `Class`</span></span>
<span><span class="fu"><a href="../reference/plot_expression_partitions.html">plot_expression_partitions</a></span><span class="op">(</span><span class="va">exp_partitions</span>, group_by <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span></span></code></pre></div>
<p><img src="HybridExpress_files/figure-html/unnamed-chunk-16-1.png" width="768"></p>
<p>You can also visualize the frequencies of genes in each partition
with the function <code><a href="../reference/plot_partition_frequencies.html">plot_partition_frequencies()</a></code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize frequency of genes in each partition</span></span>
<span><span class="co">## By `Category` (default)</span></span>
<span><span class="fu"><a href="../reference/plot_partition_frequencies.html">plot_partition_frequencies</a></span><span class="op">(</span><span class="va">exp_partitions</span><span class="op">)</span></span></code></pre></div>
<p><img src="HybridExpress_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## By `Class`</span></span>
<span><span class="fu"><a href="../reference/plot_partition_frequencies.html">plot_partition_frequencies</a></span><span class="op">(</span><span class="va">exp_partitions</span>, group_by <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span></span></code></pre></div>
<p><img src="HybridExpress_files/figure-html/unnamed-chunk-17-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="overrepresentation-analysis-of-functional-terms">Overrepresentation analysis of functional terms<a class="anchor" aria-label="anchor" href="#overrepresentation-analysis-of-functional-terms"></a>
</h2>
<p>Lastly, you’d want to explore whether gene sets of interest (e.g.,
up-regulated genes in each contrast) are enriched in any particular GO
term, pathway, protein domain, etc. For that, you will use the function
<code><a href="../reference/ora.html">ora()</a></code>, which performs overrepresentation analysis for a
gene set given a data frame of functional annotation for each gene.</p>
<p>Here, we will use an example data set with GO annotation for <em>C.
reinhardtii</em> genes. This data set illustrates how the annotation
data frame must be shaped: gene ID in the first column, and functional
annotations in other columns.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load example functional annotation (GO terms)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">go_chlamy</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">go_chlamy</span><span class="op">)</span></span>
<span><span class="co">#&gt;            gene                                           GO</span></span>
<span><span class="co">#&gt; 1 Cre01.g000050                                         &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2 Cre01.g000100                                         &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 3 Cre01.g000150                                     membrane</span></span>
<span><span class="co">#&gt; 4 Cre01.g000150                          metal ion transport</span></span>
<span><span class="co">#&gt; 5 Cre01.g000150 metal ion transmembrane transporter activity</span></span>
<span><span class="co">#&gt; 6 Cre01.g000150                      transmembrane transport</span></span></code></pre></div>
<p>To demonstrate the usage of <code><a href="../reference/ora.html">ora()</a></code>, let’s check if we can
find enrichment for any GO term among genes assigned to class “ADD”.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a vector of genes in class "ADD"</span></span>
<span><span class="va">genes_add</span> <span class="op">&lt;-</span> <span class="va">exp_partitions</span><span class="op">$</span><span class="va">Gene</span><span class="op">[</span><span class="va">exp_partitions</span><span class="op">$</span><span class="va">Class</span> <span class="op">==</span> <span class="st">"ADD"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">genes_add</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Cre01.g003650" "Cre01.g005150" "Cre01.g008600" "Cre01.g013500"</span></span>
<span><span class="co">#&gt; [5] "Cre01.g034850" "Cre01.g800005"</span></span>
<span></span>
<span><span class="co"># Get background genes - genes in the count matrix</span></span>
<span><span class="va">bg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform overrepresentation analysis</span></span>
<span><span class="va">ora_add</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ora.html">ora</a></span><span class="op">(</span><span class="va">genes_add</span>, <span class="va">go_chlamy</span>, background <span class="op">=</span> <span class="va">bg</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ora_add</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                   term genes all         pval</span></span>
<span><span class="co">#&gt; 20                      aminoacyl-tRNA ligase activity    14  33 3.811521e-05</span></span>
<span><span class="co">#&gt; 84                                           cytoplasm    43 151 7.379762e-07</span></span>
<span><span class="co">#&gt; 97                                     DNA replication    17  37 1.471956e-06</span></span>
<span><span class="co">#&gt; 114                             endopeptidase activity     6  10 7.400151e-04</span></span>
<span><span class="co">#&gt; 115                              endoplasmic reticulum     9  17 1.193854e-04</span></span>
<span><span class="co">#&gt; 118 eukaryotic translation initiation factor 3 complex     7   9 2.144087e-05</span></span>
<span><span class="co">#&gt;             padj category</span></span>
<span><span class="co">#&gt; 20  1.766005e-03       GO</span></span>
<span><span class="co">#&gt; 84  7.693402e-05       GO</span></span>
<span><span class="co">#&gt; 97  1.227611e-04       GO</span></span>
<span><span class="co">#&gt; 114 2.204188e-02       GO</span></span>
<span><span class="co">#&gt; 115 4.978369e-03       GO</span></span>
<span><span class="co">#&gt; 118 1.277263e-03       GO</span></span></code></pre></div>
<div class="section level3">
<h3 id="example-1-overrepresentation-analyses-for-all-expression-based-classes">Example 1: overrepresentation analyses for all expression-based
classes<a class="anchor" aria-label="anchor" href="#example-1-overrepresentation-analyses-for-all-expression-based-classes"></a>
</h3>
<p>In the example above, we performed an overrepresentation analysis in
genes associated with class “ADD”. What if we wanted to do the same for
<em>all</em> classes?</p>
<p>In that case, you could run <code><a href="../reference/ora.html">ora()</a></code> multiple times by
looping over each class. In details, you would do the following:</p>
<ol style="list-style-type: decimal">
<li>For each class, create a vector of genes associated with it;</li>
<li>Run <code><a href="../reference/ora.html">ora()</a></code> to get a data frame with ORA results.</li>
</ol>
<p>Below you can find an example on how to do it using
<code><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply()</a></code>. Results for each class will be stored in elements
of a list object.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a list of genes in each class</span></span>
<span><span class="va">genes_by_class</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">exp_partitions</span><span class="op">$</span><span class="va">Gene</span>, <span class="va">exp_partitions</span><span class="op">$</span><span class="va">Class</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">genes_by_class</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "UP"     "DOWN"   "ADD"    "ELD_P1" "ELD_P2"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">genes_by_class</span><span class="op">$</span><span class="va">UP</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Cre01.g029250" "Cre01.g034380" "Cre01.g036700" "Cre01.g049400"</span></span>
<span><span class="co">#&gt; [5] "Cre02.g085500" "Cre02.g087150"</span></span>
<span></span>
<span><span class="co"># Iterate through each class and perform ORA</span></span>
<span><span class="va">ora_classes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>    <span class="va">genes_by_class</span>, <span class="co"># list through which we will iterate</span></span>
<span>    <span class="va">ora</span>, <span class="co"># function we will apply to each element of the list</span></span>
<span>    annotation <span class="op">=</span> <span class="va">go_chlamy</span>, background <span class="op">=</span> <span class="va">bg</span> <span class="co"># additional arguments to function</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect output</span></span>
<span><span class="va">ora_classes</span></span>
<span><span class="co">#&gt; $UP</span></span>
<span><span class="co">#&gt;                                    term genes all         pval         padj</span></span>
<span><span class="co">#&gt; 29              ATP hydrolysis activity    21 119 3.380614e-04 0.0469905310</span></span>
<span><span class="co">#&gt; 109                      dynein complex     8  13 1.237752e-06 0.0005161427</span></span>
<span><span class="co">#&gt; 127 flavin adenine dinucleotide binding    11  42 2.857201e-04 0.0469905310</span></span>
<span><span class="co">#&gt;     category</span></span>
<span><span class="co">#&gt; 29        GO</span></span>
<span><span class="co">#&gt; 109       GO</span></span>
<span><span class="co">#&gt; 127       GO</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DOWN</span></span>
<span><span class="co">#&gt;                                  term genes all         pval         padj</span></span>
<span><span class="co">#&gt; 242           oxidoreductase activity    21 274 2.931239e-04 2.444654e-02</span></span>
<span><span class="co">#&gt; 271                    photosynthesis    13  58 3.656759e-08 7.624343e-06</span></span>
<span><span class="co">#&gt; 272  photosynthesis, light harvesting    18  24 1.638332e-22 6.831844e-20</span></span>
<span><span class="co">#&gt; 276                     photosystem I     5  15 8.644389e-05 9.011776e-03</span></span>
<span><span class="co">#&gt; 376 tetrapyrrole biosynthetic process     5   8 1.952599e-06 2.714112e-04</span></span>
<span><span class="co">#&gt;     category</span></span>
<span><span class="co">#&gt; 242       GO</span></span>
<span><span class="co">#&gt; 271       GO</span></span>
<span><span class="co">#&gt; 272       GO</span></span>
<span><span class="co">#&gt; 276       GO</span></span>
<span><span class="co">#&gt; 376       GO</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ADD</span></span>
<span><span class="co">#&gt;                                                                                      term</span></span>
<span><span class="co">#&gt; 20                                                         aminoacyl-tRNA ligase activity</span></span>
<span><span class="co">#&gt; 84                                                                              cytoplasm</span></span>
<span><span class="co">#&gt; 97                                                                        DNA replication</span></span>
<span><span class="co">#&gt; 114                                                                endopeptidase activity</span></span>
<span><span class="co">#&gt; 115                                                                 endoplasmic reticulum</span></span>
<span><span class="co">#&gt; 118                                    eukaryotic translation initiation factor 3 complex</span></span>
<span><span class="co">#&gt; 165                                                       intracellular protein transport</span></span>
<span><span class="co">#&gt; 227                                                                          nuclear pore</span></span>
<span><span class="co">#&gt; 250 oxidoreductase activity, acting on the CH-OH group of donors, NAD or NADP as acceptor</span></span>
<span><span class="co">#&gt; 284                                                                     prefoldin complex</span></span>
<span><span class="co">#&gt; 287                                                               proteasome core complex</span></span>
<span><span class="co">#&gt; 293                                                                       protein folding</span></span>
<span><span class="co">#&gt; 294                                                                 protein glycosylation</span></span>
<span><span class="co">#&gt; 304                                                                     protein refolding</span></span>
<span><span class="co">#&gt; 390                                                translation initiation factor activity</span></span>
<span><span class="co">#&gt; 410                                                              unfolded protein binding</span></span>
<span><span class="co">#&gt; 414                                                            vesicle-mediated transport</span></span>
<span><span class="co">#&gt;     genes all         pval         padj category</span></span>
<span><span class="co">#&gt; 20     14  33 3.811521e-05 1.766005e-03       GO</span></span>
<span><span class="co">#&gt; 84     43 151 7.379762e-07 7.693402e-05       GO</span></span>
<span><span class="co">#&gt; 97     17  37 1.471956e-06 1.227611e-04       GO</span></span>
<span><span class="co">#&gt; 114     6  10 7.400151e-04 2.204188e-02       GO</span></span>
<span><span class="co">#&gt; 115     9  17 1.193854e-04 4.978369e-03       GO</span></span>
<span><span class="co">#&gt; 118     7   9 2.144087e-05 1.277263e-03       GO</span></span>
<span><span class="co">#&gt; 165    25  73 4.442015e-06 3.087201e-04       GO</span></span>
<span><span class="co">#&gt; 227     6   9 3.339022e-04 1.160310e-02       GO</span></span>
<span><span class="co">#&gt; 250    10  27 1.739601e-03 4.533836e-02       GO</span></span>
<span><span class="co">#&gt; 284     4   4 3.203862e-04 1.160310e-02       GO</span></span>
<span><span class="co">#&gt; 287     9  15 3.165895e-05 1.650223e-03       GO</span></span>
<span><span class="co">#&gt; 293    29  81 2.663812e-07 3.702698e-05       GO</span></span>
<span><span class="co">#&gt; 294     6  10 7.400151e-04 2.204188e-02       GO</span></span>
<span><span class="co">#&gt; 304    10  10 1.809949e-09 7.547489e-07       GO</span></span>
<span><span class="co">#&gt; 390     9  21 8.722652e-04 2.424897e-02       GO</span></span>
<span><span class="co">#&gt; 410    19  38 6.478683e-08 1.350805e-05       GO</span></span>
<span><span class="co">#&gt; 414    13  41 1.955735e-03 4.797302e-02       GO</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ELD_P1</span></span>
<span><span class="co">#&gt;                term genes all         pval        padj category</span></span>
<span><span class="co">#&gt; 38          binding    88 208 2.494608e-06 0.001040251       GO</span></span>
<span><span class="co">#&gt; 239         nucleus    77 193 1.141172e-04 0.023793431       GO</span></span>
<span><span class="co">#&gt; 345 rRNA processing    15  24 3.464517e-04 0.048156786       GO</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ELD_P2</span></span>
<span><span class="co">#&gt;                                                                term genes all</span></span>
<span><span class="co">#&gt; 28                             aspartic-type endopeptidase activity     6   8</span></span>
<span><span class="co">#&gt; 170                                                iron ion binding    30  93</span></span>
<span><span class="co">#&gt; 242                                         oxidoreductase activity    71 274</span></span>
<span><span class="co">#&gt; 271                                                  photosynthesis    26  58</span></span>
<span><span class="co">#&gt; 278                                                  photosystem II    17  32</span></span>
<span><span class="co">#&gt; 320 proton-transporting two-sector ATPase complex, catalytic domain     6   8</span></span>
<span><span class="co">#&gt;             pval         padj category</span></span>
<span><span class="co">#&gt; 28  3.007465e-04 2.090188e-02       GO</span></span>
<span><span class="co">#&gt; 170 4.577693e-05 4.772245e-03       GO</span></span>
<span><span class="co">#&gt; 242 6.069186e-06 8.436168e-04       GO</span></span>
<span><span class="co">#&gt; 271 1.191954e-07 4.970448e-05       GO</span></span>
<span><span class="co">#&gt; 278 9.602908e-07 2.002206e-04       GO</span></span>
<span><span class="co">#&gt; 320 3.007465e-04 2.090188e-02       GO</span></span></code></pre></div>
<p>To do the same for each expression-based category (not class), you’d
need to replace <code>Class</code> with <code>Category</code> in the
<code><a href="https://rdrr.io/r/base/split.html" class="external-link">split()</a></code> function (see example above).</p>
</div>
<div class="section level3">
<h3 id="example-2-overrepresentation-analyses-for-differentially-expressed-genes">Example 2: overrepresentation analyses for differentially expressed
genes<a class="anchor" aria-label="anchor" href="#example-2-overrepresentation-analyses-for-differentially-expressed-genes"></a>
</h3>
<p>The same you can use <code><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply()</a></code> to loop through each
expression class, you can also loop through each contrast in the list
returned by <code><a href="../reference/get_deg_list.html">get_deg_list()</a></code>, and perform ORA for up- and
down-regulated genes. Below you can find an example:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get up-regulated genes for each contrast</span></span>
<span><span class="va">up_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">deg_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">x</span><span class="op">$</span><span class="va">log2FoldChange</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">up_genes</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "P2_vs_P1"        "F1_vs_P1"        "F1_vs_P2"        "F1_vs_midparent"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">up_genes</span><span class="op">$</span><span class="va">F1_vs_P1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Cre01.g000750" "Cre01.g000900" "Cre01.g001200" "Cre01.g002750"</span></span>
<span><span class="co">#&gt; [5] "Cre01.g003524" "Cre01.g003650"</span></span>
<span></span>
<span><span class="co"># Perform ORA</span></span>
<span><span class="va">ora_up</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>    <span class="va">up_genes</span>,</span>
<span>    <span class="va">ora</span>,</span>
<span>    annotation <span class="op">=</span> <span class="va">go_chlamy</span>, background <span class="op">=</span> <span class="va">bg</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ora_up</span></span>
<span><span class="co">#&gt; $P2_vs_P1</span></span>
<span><span class="co">#&gt; [1] term     genes    all      pval     padj     category</span></span>
<span><span class="co">#&gt; &lt;0 rows&gt; (or 0-length row.names)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $F1_vs_P1</span></span>
<span><span class="co">#&gt;               term genes all         pval        padj category</span></span>
<span><span class="co">#&gt; 109 dynein complex     8  13 2.049704e-05 0.008547264       GO</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $F1_vs_P2</span></span>
<span><span class="co">#&gt;                                              term genes all         pval</span></span>
<span><span class="co">#&gt; 38                                        binding   138 208 1.214715e-06</span></span>
<span><span class="co">#&gt; 96                                     DNA repair    56  83 9.520576e-04</span></span>
<span><span class="co">#&gt; 144                                   GTP binding    72 111 1.095054e-03</span></span>
<span><span class="co">#&gt; 169                                 ion transport    33  45 1.225270e-03</span></span>
<span><span class="co">#&gt; 209                     microtubule-based process    10  10 9.792020e-04</span></span>
<span><span class="co">#&gt; 239                                       nucleus   125 193 2.252487e-05</span></span>
<span><span class="co">#&gt; 303                        protein polymerization    10  10 9.792020e-04</span></span>
<span><span class="co">#&gt; 304                             protein refolding    10  10 9.792020e-04</span></span>
<span><span class="co">#&gt; 307 protein serine/threonine phosphatase activity    20  24 7.724253e-04</span></span>
<span><span class="co">#&gt; 338                                   RNA binding   101 157 1.948736e-04</span></span>
<span><span class="co">#&gt; 384                  transcription, DNA-templated    28  36 5.951883e-04</span></span>
<span><span class="co">#&gt;            padj category</span></span>
<span><span class="co">#&gt; 38  0.000506536       GO</span></span>
<span><span class="co">#&gt; 96  0.045369693       GO</span></span>
<span><span class="co">#&gt; 144 0.045663753       GO</span></span>
<span><span class="co">#&gt; 169 0.046448881       GO</span></span>
<span><span class="co">#&gt; 209 0.045369693       GO</span></span>
<span><span class="co">#&gt; 239 0.004696435       GO</span></span>
<span><span class="co">#&gt; 303 0.045369693       GO</span></span>
<span><span class="co">#&gt; 304 0.045369693       GO</span></span>
<span><span class="co">#&gt; 307 0.045369693       GO</span></span>
<span><span class="co">#&gt; 338 0.027087430       GO</span></span>
<span><span class="co">#&gt; 384 0.045369693       GO</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $F1_vs_midparent</span></span>
<span><span class="co">#&gt;                                              term genes all         pval</span></span>
<span><span class="co">#&gt; 80         cyclic nucleotide biosynthetic process    38 108 1.983329e-04</span></span>
<span><span class="co">#&gt; 109                                dynein complex     9  13 1.789139e-04</span></span>
<span><span class="co">#&gt; 166             intracellular signal transduction    38 110 3.036659e-04</span></span>
<span><span class="co">#&gt; 168                          ion channel activity    20  48 5.611617e-04</span></span>
<span><span class="co">#&gt; 169                                 ion transport    21  45 5.765412e-05</span></span>
<span><span class="co">#&gt; 290                     protein dephosphorylation    21  44 3.792684e-05</span></span>
<span><span class="co">#&gt; 307 protein serine/threonine phosphatase activity    15  24 7.437911e-06</span></span>
<span><span class="co">#&gt;            padj category</span></span>
<span><span class="co">#&gt; 80  0.016540962       GO</span></span>
<span><span class="co">#&gt; 109 0.016540962       GO</span></span>
<span><span class="co">#&gt; 166 0.021104781       GO</span></span>
<span><span class="co">#&gt; 168 0.033429202       GO</span></span>
<span><span class="co">#&gt; 169 0.008013923       GO</span></span>
<span><span class="co">#&gt; 290 0.007907747       GO</span></span>
<span><span class="co">#&gt; 307 0.003101609       GO</span></span></code></pre></div>
<p>Likewise, for down-regulated genes, you need to replace the
<code>&gt;</code> symbol with a <code>&lt;</code> symbol in the
anonymous function to subset rows.</p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="faq">FAQ<a class="anchor" aria-label="anchor" href="#faq"></a>
</h2>
<ol style="list-style-type: decimal">
<li>How do I create a <code>SummarizedExperiment</code> object?</li>
</ol>
<p>A <code>SummarizedExperiment</code> is a data structure (an S4 class)
that can be used to store, in a single object, the following
elements:</p>
<ol style="list-style-type: decimal">
<li>
<strong>assay:</strong> A quantitative matrix with features in rows
and samples in columns. In the context of <em><a href="https://github.com/HybridExpress" class="external-link">HybridExpress</a></em>, this
would be a gene expression matrix (in counts) with genes in rows and
samples in columns.</li>
<li>
<strong>colData:</strong> A data frame of sample metadata with
samples in rows and variables that describe samples (e.g., tissue,
treatment, and other covariates) in columns.</li>
<li>
<strong>rowData:</strong> A data frame of gene metadata with genes
in rows and variables that describe genes (e.g., chromosome name,
alternative IDs, functional information, etc) in columns.</li>
</ol>
<p>For this package, you must have an <em>assay</em> containing the
count matrix and a <em>colData</em> slot with sample metadata.
<em>rowData</em> can be present, but it is not required.</p>
<p>To demonstrate how to create a <code>SummarizedExperiment</code>
object, we will extract the <em>assay</em> and <em>colData</em> from the
example object <code>se_chlamy</code> that comes with this package.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get count matrix</span></span>
<span><span class="va">count_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_chlamy</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">count_matrix</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 S1  S2   S3   S4  S5   S6   S7   S8   S9 S10 S11  S12  S13  S14</span></span>
<span><span class="co">#&gt; Cre01.g000050   31  21   26   33  19   48   17    6   13   6  10    7   31   22</span></span>
<span><span class="co">#&gt; Cre01.g000150   50  29   35   99  11   58   51   46   41  33  28   14   53   25</span></span>
<span><span class="co">#&gt; Cre01.g000200   36  26   24   29  17   36   14   15   12   8  14    7   25   24</span></span>
<span><span class="co">#&gt; Cre01.g000250  440 272  394  332 283  585  272  274  255 160 225  235  405  391</span></span>
<span><span class="co">#&gt; Cre01.g000300 1242 839 1216 1251 811 1785 1341 1306 1122 877 844 1082 1704 1739</span></span>
<span><span class="co">#&gt; Cre01.g000350  412 264  294  336 252  478  233  221  195 155 190  201  299  272</span></span>
<span><span class="co">#&gt;                S15  S16  S17  S18</span></span>
<span><span class="co">#&gt; Cre01.g000050   29   22   21   16</span></span>
<span><span class="co">#&gt; Cre01.g000150   37   17   24   21</span></span>
<span><span class="co">#&gt; Cre01.g000200   24   26   18   21</span></span>
<span><span class="co">#&gt; Cre01.g000250  358  339  340  332</span></span>
<span><span class="co">#&gt; Cre01.g000300 1524 1720 1517 1243</span></span>
<span><span class="co">#&gt; Cre01.g000350  276  324  246  275</span></span>
<span></span>
<span><span class="co"># Get colData (data frame of sample metadata)</span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se_chlamy</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 2 columns</span></span>
<span><span class="co">#&gt;         Ploidy Generation</span></span>
<span><span class="co">#&gt;    &lt;character&gt;   &lt;factor&gt;</span></span>
<span><span class="co">#&gt; S1     diploid         P1</span></span>
<span><span class="co">#&gt; S2     diploid         P1</span></span>
<span><span class="co">#&gt; S3     diploid         P1</span></span>
<span><span class="co">#&gt; S4     diploid         P1</span></span>
<span><span class="co">#&gt; S5     diploid         P1</span></span>
<span><span class="co">#&gt; S6     diploid         P1</span></span></code></pre></div>
<p>Once you have these two objects, you can create a
<code>SummarizedExperiment</code> object with:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a SummarizedExperiment object</span></span>
<span><span class="va">new_se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>    assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_matrix</span><span class="op">)</span>,</span>
<span>    colData <span class="op">=</span> <span class="va">coldata</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">new_se</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 13058 18 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(13058): Cre01.g000050 Cre01.g000150 ... ERCC-00170 ERCC-00171</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(18): S1 S2 ... S17 S18</span></span>
<span><span class="co">#&gt; colData names(2): Ploidy Generation</span></span></code></pre></div>
<p>For more details on this data structure, read the vignette of the
<em><a href="https://bioconductor.org/packages/3.19/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
package.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>This document was created under the following conditions:</p>
<pre><code><span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R Under development (unstable) (2024-02-07 r85873)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  en_US.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    en_US.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2024-02-15</span></span>
<span><span class="co">#&gt;  pandoc   3.1.1 @ /usr/local/bin/ (via rmarkdown)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package             </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version  </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  abind                  1.4-5     <span style="color: #555555;">2016-07-21</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  Biobase              * 2.63.0    <span style="color: #555555;">2023-10-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocGenerics         * 0.49.1    <span style="color: #555555;">2023-11-01</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocManager            1.30.22   <span style="color: #555555;">2023-08-08</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  BiocParallel           1.37.0    <span style="color: #555555;">2023-10-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocStyle            * 2.31.0    <span style="color: #555555;">2023-10-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  bitops                 1.0-7     <span style="color: #555555;">2021-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  bookdown               0.37      <span style="color: #555555;">2023-12-01</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  bslib                  0.6.1     <span style="color: #555555;">2023-11-28</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  cachem                 1.0.8     <span style="color: #555555;">2023-05-01</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  circlize               0.4.15    <span style="color: #555555;">2022-05-10</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  cli                    3.6.2     <span style="color: #555555;">2023-12-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  clue                   0.3-65    <span style="color: #555555;">2023-09-23</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  cluster                2.1.6     <span style="color: #555555;">2023-12-01</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  codetools              0.2-19    <span style="color: #555555;">2023-02-01</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  colorspace             2.1-0     <span style="color: #555555;">2023-01-23</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  ComplexHeatmap         2.19.0    <span style="color: #555555;">2023-10-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  crayon                 1.5.2     <span style="color: #555555;">2022-09-29</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  DelayedArray           0.29.1    <span style="color: #555555;">2024-01-31</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  desc                   1.4.3     <span style="color: #555555;">2023-12-10</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  DESeq2                 1.43.2    <span style="color: #555555;">2024-02-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  digest                 0.6.34    <span style="color: #555555;">2024-01-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  doParallel             1.0.17    <span style="color: #555555;">2022-02-07</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  dplyr                  1.1.4     <span style="color: #555555;">2023-11-17</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  evaluate               0.23      <span style="color: #555555;">2023-11-01</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  fansi                  1.0.6     <span style="color: #555555;">2023-12-08</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  farver                 2.1.1     <span style="color: #555555;">2022-07-06</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  fastmap                1.1.1     <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  foreach                1.5.2     <span style="color: #555555;">2022-02-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  fs                     1.6.3     <span style="color: #555555;">2023-07-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  generics               0.1.3     <span style="color: #555555;">2022-07-05</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDb         * 1.39.6    <span style="color: #555555;">2024-02-08</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDbData       1.2.11    <span style="color: #555555;">2024-02-04</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  GenomicRanges        * 1.55.3    <span style="color: #555555;">2024-02-13</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  GetoptLong             1.0.5     <span style="color: #555555;">2020-12-15</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  ggplot2                3.4.4     <span style="color: #555555;">2023-10-12</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  GlobalOptions          0.1.2     <span style="color: #555555;">2020-06-10</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  glue                   1.7.0     <span style="color: #555555;">2024-01-09</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  gtable                 0.3.4     <span style="color: #555555;">2023-08-21</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  highr                  0.10      <span style="color: #555555;">2022-12-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  htmltools              0.5.7     <span style="color: #555555;">2023-11-03</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  HybridExpress        * 0.99.0    <span style="color: #555555;">2024-02-15</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  IRanges              * 2.37.1    <span style="color: #555555;">2024-01-19</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  iterators              1.0.14    <span style="color: #555555;">2022-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  jquerylib              0.1.4     <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  jsonlite               1.8.8     <span style="color: #555555;">2023-12-04</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  knitr                  1.45      <span style="color: #555555;">2023-10-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  labeling               0.4.3     <span style="color: #555555;">2023-08-29</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  lattice                0.22-5    <span style="color: #555555;">2023-10-24</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  lifecycle              1.0.4     <span style="color: #555555;">2023-11-07</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  locfit                 1.5-9.8   <span style="color: #555555;">2023-06-11</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  magrittr               2.0.3     <span style="color: #555555;">2022-03-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  Matrix                 1.6-5     <span style="color: #555555;">2024-01-11</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  MatrixGenerics       * 1.15.0    <span style="color: #555555;">2023-10-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  matrixStats          * 1.2.0     <span style="color: #555555;">2023-12-11</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  memoise                2.0.1     <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  munsell                0.5.0     <span style="color: #555555;">2018-06-12</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  patchwork              1.2.0     <span style="color: #555555;">2024-01-08</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  pillar                 1.9.0     <span style="color: #555555;">2023-03-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  pkgconfig              2.0.3     <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  pkgdown                2.0.7     <span style="color: #555555;">2022-12-14</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  png                    0.1-8     <span style="color: #555555;">2022-11-29</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  purrr                  1.0.2     <span style="color: #555555;">2023-08-10</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  R6                     2.5.1     <span style="color: #555555;">2021-08-19</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  ragg                   1.2.7     <span style="color: #555555;">2023-12-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  RColorBrewer           1.1-3     <span style="color: #555555;">2022-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  Rcpp                   1.0.12    <span style="color: #555555;">2024-01-09</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  RCurl                  1.98-1.14 <span style="color: #555555;">2024-01-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  rjson                  0.2.21    <span style="color: #555555;">2022-01-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  rlang                  1.1.3     <span style="color: #555555;">2024-01-10</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  rmarkdown              2.25      <span style="color: #555555;">2023-09-18</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  S4Arrays               1.3.3     <span style="color: #555555;">2024-01-31</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  S4Vectors            * 0.41.3    <span style="color: #555555;">2024-01-01</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  sass                   0.4.8     <span style="color: #555555;">2023-12-06</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  scales                 1.3.0     <span style="color: #555555;">2023-11-28</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  sessioninfo            1.2.2     <span style="color: #555555;">2021-12-06</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  shape                  1.4.6     <span style="color: #555555;">2021-05-19</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  SparseArray            1.3.4     <span style="color: #555555;">2024-02-11</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  stringi                1.8.3     <span style="color: #555555;">2023-12-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  stringr                1.5.1     <span style="color: #555555;">2023-11-14</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  SummarizedExperiment * 1.33.3    <span style="color: #555555;">2024-01-23</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  systemfonts            1.0.5     <span style="color: #555555;">2023-10-09</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  textshaping            0.3.7     <span style="color: #555555;">2023-10-09</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  tibble                 3.2.1     <span style="color: #555555;">2023-03-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  tidyselect             1.2.0     <span style="color: #555555;">2022-10-10</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  utf8                   1.2.4     <span style="color: #555555;">2023-10-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  vctrs                  0.6.5     <span style="color: #555555;">2023-12-01</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  withr                  3.0.0     <span style="color: #555555;">2024-01-16</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  xfun                   0.42      <span style="color: #555555;">2024-02-08</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  XVector                0.43.1    <span style="color: #555555;">2024-01-10</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor 3.19 (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  yaml                   2.3.8     <span style="color: #555555;">2023-12-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.0)</span></span></span>
<span><span class="co">#&gt;  zlibbioc               1.49.0    <span style="color: #555555;">2023-10-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /__w/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /usr/local/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /usr/local/lib/R/library</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-coate2023beyond" class="csl-entry">
Coate, Jeremy E. 2023. <span>“Beyond Transcript Concentrations:
Quantifying Polyploid Expression Responses Per Biomass, Per Genome, and
Per Cell with RNA-Seq.”</span> In <em>Polyploidy: Methods and
Protocols</em>, 227–50. Springer.
</div>
<div id="ref-rapp2009genomic" class="csl-entry">
Rapp, Ryan A, Joshua A Udall, and Jonathan F Wendel. 2009.
<span>“Genomic Expression Dominance in Allopolyploids.”</span> <em>BMC
Biology</em> 7 (1): 1–10.
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Fabricio Almeida-Silva, Lucas Prost-Boxoen, Yves Van de Peer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
